extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpAuthenticator.java

```diff
@@ -309,6 +342,39 @@ private static Uni<HttpCredentialTransport> getCredentialTransport(HttpAuthentic
         }
     }
 
+    private static void rememberAuthAttempted(RoutingContext routingContext) {
+        routingContext.put(ATTEMPT_AUTH_INVOKED, TRUE);
+    }
+
+    private static boolean isAuthMechanismSelected(RoutingContext routingContext) {
+        return routingContext.get(AUTH_MECHANISM) != null;
+    }
+
+    private static boolean requestAlreadyAuthenticated(RoutingContext event, String newAuthMechanism) {
+        return event.get(ATTEMPT_AUTH_INVOKED) == TRUE && authenticatedWithDifferentAuthMechanism(newAuthMechanism, event);
+    }
+
+    private static boolean authenticatedWithDifferentAuthMechanism(String newAuthMechanism, RoutingContext event) {
+        return !newAuthMechanism.equalsIgnoreCase(event.get(AUTH_MECHANISM));
+    }
+
+    /**
+     * Remember authentication mechanism used for authentication so that we know what mechanism has been used
+     * in case that someone tries to change the mechanism after the authentication. This way, we can be permissive
+     * when the selected mechanism is same as the one already used.
+     */
+    private static Uni<HttpCredentialTransport> rememberAuthMechScheme(HttpAuthenticationMechanism mech, RoutingContext event) {
```

Comment:
@michalvavrik Can you please clarify why is it necessary, we don't need it with multiple mechanisms with HTTP security policies

Reply:
@sberyozkin 

> @michalvavrik Can you please clarify why is it necessary,

I'll try to provide couple of answers, choose which one you like (if any :-)):

1. Analogy to `@Tenant` annotation. If someone applies this annotation, he expects the tenant from the annotation value to be used during the authentication. Which means that authentication can only happen in "after match" phase. Anything before should be illegal state (but is not, you can do that; I don't want to discuss it here, sorry, if you have questions in that regard, please open separate issue, I absolutely understand such need!). I try to prevent detect illegal state and provide information to the user because matching the HTTP request with the resource method can only dynamically during the runtime. Users can influence it, though in practice I don't believe it happens.,
2. This is safeguard. It keeps things safe and consistent. It provides users information if their intention to use the `CodeFlow` or `MTLS` etc. didn't happen because they did setup their application incorrectly. It must exist because matching the resource method with the HTTP request can only happen dynamically during the request.
3. Only reason why this must exist is that it will always be theoretically possible to perform authentication in pre-match phase. One example I can give you is custom `SecurityContext` in a pre-match JAX-RS filter or if user doesn't use value `JAX-RS` in the https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus-http-auth-permission-permissions-applies-to for endpoints annotated with `@CodeFlow` or `@Tenant`.

> we don't need it with multiple mechanisms with HTTP security policies

The `ALL` HTTP Security Policies can be run in pre-match phase. Before the resource method has been matched with the HTTP request.

Reply:
@michalvavrik I think it makes sense